### PR TITLE
fix #8980 - improve horizontal placement of tie endpoints

### LINKED ISSUE #8980: [MU4 Issue] Improve horizontal position of chord tie endpoints
- add an option to start ties after augmentation dots, where present;
- on left, take into account bounding boxes of all elements of starting chord: flags, ledger lines, stems, offset noteheads, dots (if the aforementioned option has been chosen), etc.;
- on the right, take into account bounding boxes of all elements of the ending chord, similarly (dots and flags will not be relevant).

It appears that horizontal placement is determined first, and that this is dependent on the available space, and therefore can change when the music is respaced (at least for now); the horizontal offsets will need to be recalculated then too.

Some approximate before-and-afters:

chord-ties

This is a first step towards addressing https://musescore.org/en/node/45576 and many of the issues linked there. (Horizontal position, curvature, behaviour for single notes etc. all to follow.)

### PR DESCRIPTION
**Pull request by asattely**
addresses #8980 Improve horizontal position of chord tie endpoints

Tie endpoints now adjust for offset noteheads, ledger lines, flags, and dots. This PR does not deal with the vertical placement of these ties; that will be a different issue popping up in the near future.

### CODE REVIEW

**Comment by asattely** [comment]
fixed some codestyle stuff and a stupid mistake

File: src/engraving/libmscore/tie.cpp [hunk 1]
```diff
@@ -494,17 +497,66 @@ void Tie::slurPos(SlurPos* sp)
...
+
+        // adjust for flags (hooks)
+        if (sc->hook() && sc->up()) {
+            if (y1 < sc->hook()->pos().y() + sc->hook()->bbox().height() + (offsetMargin * 2)) {
+                xo = qMax(xo, sc->hook()->pos().x() + sc->hook()->bbox().width());
+            }
+        }
+
+        // adjust for ledger lines
+        if (startNote()->line() >= 8 || startNote()->line() <= 0) {
```
Comment (cbjeukendrup) [inline src/engraving/libmscore/tie.cpp:535]: This assumes that the notes are on a normal 5 line staff. You actually need check `sc->staff()->staffTypeForElement(sc)->lines()` to get the real number of lines.

File: src/engraving/libmscore/tie.cpp [hunk 2]
```diff
@@ -518,23 +570,53 @@ void Tie::slurPos(SlurPos* sp)
...
-    if ((ec->notes().size() > 1) || (ec->stem() && !ec->up() && !_up)) {
-        xo = endNote()->x() - hw * 0.12;
+    if (ec->notes().size() > 1) {
+        x2 = endNote()->x();
+        qreal xo = 0.0;
+
+        // ADJUST FOR COLLISIONS ----------------
+
+        // adjust for ledger lines
+        if (endNote()->line() >= 8 || endNote()->line() <= 0) {
```
Comment (cbjeukendrup) [inline src/engraving/libmscore/tie.cpp:580]: Same here.

Comment (asattely) [reply src/engraving/libmscore/tie.cpp:580]: Nice catch! I'll adjust that tomorrow.

**Comment by Jojo-Schmitz** [comment]
rebase needed

**Comment by Jojo-Schmitz** [comment]
That wasn't a proper rebase...

**Comment by asattely** [comment]
rebased, and also fixed a small problem with ties being adjusted for dots on offset notes

**Review by oktophonie — APPROVED** [review]

**Review by RomanPudashkin — APPROVED** [review]

**Pull request merged**